### DISCUSSION TITLE
Official grammer/spec

### DISCUSSION THREAD

**Discussion by happenslol** [microsoft/typespec#3088] [General]
I'm writing a treesitter grammar over at [happenslol/tree-sitter-typespec](https://github.com/happenslol/tree-sitter-typespec), so I can use typespec in neovim. I can currently parse the example on the typespec homepage, but before I dive in further, I wanted to ask: Is there an official grammar or a spec for the language?

I've pulled things like identifier and string literal definitions from the typescript grammar, since from my understanding this started out as a typescript library. I could probably go through all the example code and tutorials and pull out all of the syntax, but it would be much easier if there was an official spec I could use.

Thanks!

**Comment by timotheeguerin** [comment]
Our offical grammar spec can be found here https://github.com/microsoft/typespec/blob/main/packages/spec/src/spec.emu.html, you can download and open the file in a browser to get the grammar UI

I know we have an item about directives (e.g. `#suppress`) not being described correctly in the grammar but otherwise it should be up  to date.

**Reply by happenslol** [reply]
Awesome, thanks for the quick response!

**Reply by timotheeguerin** [reply]
Also it is awesome to see that getting integrated, I have no idea what are the requirements to integrate like that with neovim but we also have a tmlanguage and language server you can make use of

**Reply by happenslol** [reply]
There's a few parts needed to integrate a language completely into neovim:

- A grammar for highlighting/indentation etc. It's possible to create a "regular" regex based grammar, but treesitter seems to be preferred these days (and is much easier from my experience). It's my first time writing a TS grammar but I should be able to get it done relatively quickly using the grammar and tmlanguage you provided. I'll try to submit this to the official nvim-treesitter repo as well, if nobody else is faster than me.
- LSP integration. This can be set up manually, but an entry in [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig) makes this a one-liner. I've already got this locally, will submit it ASAP.
- Formatting. From what I can tell the LSP doesn't format itself, but `tsp format` does the job. While it would be nicer to have the lsp do formatting directly, [none-ls](https://github.com/nvimtools/none-ls.nvim) is basically a virtual lsp that can call out to other executables and make the config a one-liner as well. I'll submit a builtin for typespec there as well.

Really hyped for this project, I hope I can help adoption by improving the neovim DX!

**Reply by timotheeguerin** [reply]
The language server does have the formatting capability setup(whole file, we don’t support partial formatting yet)
Let me know if there is an issue

**Reply by happenslol** [reply]
Ah right, then maybe I've just misconfigured the LSP capabilities in my config. Thanks for the heads up, that makes it easier!

**Reply by happenslol** [reply]
I'm pretty much done with the grammar itself. I have 2 further questions:

- Projection statements are still in the spec you linked, but they are not in the statement list and can never appear anywhere from what I can see. I saw that they are deprecated anyways, is it fine to leave them out?
- Is there something like a test document which you use internally to validate your grammar or parser? I've written some tests myself, but it would be nice to be able to run against something official to make sure the grammar stays compliant.

All that's left to do for me is to write queries for syntax highlighting and scopes. I'll give you a heads up when I go for inclusion in the nvim-treesitter repository.

**Reply by timotheeguerin** [reply]
Awesome to hear the progress!

- projection: yes do not spend time on this, they are a very experimental feature that is not at all in a good state today
- for the test we have this file which check we tokenized things correctly both with `tmgrammar` and semantic highlighting https://github.com/microsoft/typespec/blob/main/packages/compiler/test/server/colorization.test.ts  you'll most likely have to tweak things but if you find that this is something you can reuse I am open to refactoring our test a little bit to make it easier to plugin for other systems and providing it as a standard test suite

**Reply by happenslol** [reply]
Done with the grammar and I'm using it for highlighting myself. Haven't had the time to open a PR for `nvim-treesitter` to get it added as an official grammar yet, but I think I'll get to that soon. I'm also still waiting on the `typespec` filetype being upstreamed into neovim, which I've already submitted, so it should be in the next release.

As for your test files, I think it'd be pretty hard to integrate that into my tests. Tree-sitter itself offers very easy snapshot-like tests which I'm currently using to make sure (some of) your example documents from the website are parsed correctly, but the syntax tree is very different to what your tests expect. It would be nice to have something automated in the long run though. You can have a look at my current tests [here](https://github.com/happenslol/tree-sitter-typespec/blob/main/test/corpus/statements.txt).

**Reply by happenslol** [reply]
The typespec grammar has been [merged into nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter/pull/6624)! This means with the next release, there is now builtin syntax highlighting for typespec.

**Reply by timotheeguerin** [reply]
awesome to hear!! thanks for driving this!
